File: armorpaint/assets/plugins/hello_node_brush.js

let plugin = plugin_create();

let category_name = "My Nodes";
let node_name = "Hello World";
let node_type = "HELLO_WORLD";

// Create new node category
let node_list = [
	{
		id: 0,
		name: node_name,
		type: node_type,
		x: 0,
		y: 0,
		color: 0xffb34f5a,
		inputs: [
			{
				id: 0,
				node_id: 0,
				name: "Scale",
				type: "VALUE",
				color: 0xffa1a1a1,
				default_value__f32: [1.0],
				min__f32: 0.0,
				max__f32: 5.0,
				precision__f32: 100.0,
				display: 0
			}
		],
		outputs: [
			{
				id: 0,
				node_id: 0,
				name: "Value",
				type: "VALUE",
				color: 0xffc7c729,
				default_value__f32: [1.0],
				min__f32: 0.0,
				max__f32: 5.0,
				precision__f32: 100.0,
				display: 0
			}
		],
		buttons: [],
		width: 0
	}
];

nodes_brush_category_add(category_name, armpack_encode(node_list));

// Brush node
parser_logic_custom_nodes_set(node_type, function(node, from) {
	return Math.sin(sys_time() * node.inputs[0].get(0));
});

// Cleanup
plugin_notify_on_delete(plugin, function() {
	parser_logic_custom_nodes_delete(node_type);
	nodes_brush_category_remove(category_name);
});

File: paint/assets/plugins/hello_node_brush.c
#include "global.h"

void *plugin;
char *category_name = "My Nodes";
char *node_name = "Hello World";
char *node_type = "HELLO_WORLD";

float custom_node(logic_node_t *node, int from) {
	return sin(sys_time() * node->inputs->buffer[0]->get(0));
}

void on_delete() {
	plugin_brush_custom_nodes_remove(node_type);
	plugin_brush_category_remove(category_name);
}

void main() {
	plugin = plugin_create();
	gc_root(plugin);

	// Create new node category
	any_array_t *node_list = any_array_create(0);
	// ui_node_t *n = gc_alloc(sizeof(ui_node_t));
	ui_node_t *n = gc_alloc(72);
	any_array_push(node_list, n);

	n->id = 0;
	n->name = node_name;
	n->type = node_type;
	n->x = 0;
	n->y = 0;
	n->color = 0xffb34f5a;

	n->inputs = any_array_create(0);
	// ui_node_socket_t *s = gc_alloc(sizeof(ui_node_socket_t));
	ui_node_socket_t *s = gc_alloc(56);
	any_array_push(n->inputs, s);
	s->id = 0;
	s->node_id = 0;
	s->name = "Scale";
	s->type = "VALUE";
	s->color = 0xffa1a1a1;
	s->default_value = f32_array_create_x(1.0);
	s->min = 0.0;
	s->max = 5.0;
	s->precision = 100.0;
	s->display = 0;

	n->outputs = any_array_create(0);
	// s = gc_alloc(sizeof(ui_node_socket_t));
	s = gc_alloc(56);
	any_array_push(n->outputs, s);
	s->id = 0;
	s->node_id = 0;
	s->name = "Value";
	s->type = "VALUE";
	s->color = 0xffc7c729;
	s->default_value = f32_array_create_x(1.0);
	s->min = 0.0;
	s->max = 5.0;
	s->precision = 100.0;
	s->display = 0;

	n->buttons = any_array_create(0);
	n->width = 0;
	n->flags = 0;

	plugin_brush_category_add(category_name, node_list);

	// Brush node
	plugin_brush_custom_nodes_set(node_type, custom_node);

	// Cleanup
	plugin_notify_on_delete(plugin, on_delete);
}
